pycbc/types/array_cpu.pyx

```diff
@@ -113,27 +84,47 @@ def weighted_inner(self, other, weight):
 
     return _np.sum(self.data.conj() * other / weight, dtype=acum_dtype)
 
+ctypedef fused REALTYPE:
+    float
+    double
 
-inner_code = """
-double value = 0;
-
-#pragma omp parallel for reduction(+:value)
-for (int i=0; i<N; i++){
-    float val = x[i] * y[i];
-    value += val;
-}
-total[0] = value;
-"""
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def inner_real(numpy.ndarray [REALTYPE, ndim = 1] a, numpy.ndarray [REALTYPE, ndim = 1] b):
+    cdef double total = 0
+    cdef unsigned int xmax = a.shape[0]
+    cdef unsigned int i
+    
+    cdef REALTYPE* x = &a[0]
+    cdef REALTYPE* y = &b[0]
 
+    for i in range(xmax):
+        total += x[i] * y[i]
+    return total
+    
+ctypedef fused COMPLEXTYPE:
+    float complex
+    double complex
+
+def abs_arg_max_complex(numpy.ndarray [COMPLEXTYPE, ndim=1] a):
+    cdef unsigned int xmax = a.shape[0]
+    cdef double mag
+    cdef double magmax = 0
+    cdef unsigned int idx = 0
```

Comment:
Should `mag` and `magmax` be declared double, or can they also be `REALTYPE`? So that you don't impose a conversion every time you do the comparison if you're using single precision?

Reply:
Actually, I think I couldn't figure out how to do that, so I went with higher precision. What is happening is that it is generating a kernel for every combination of COMPLEXTYPE, since that is in the interface. It doesn't do so for REALTYPE, unless there were an argument, I think....

Reply:
Maybe to explain better. It has to know how to resolve fused types based on the signature. It can't really know how to do so for REALTYPE, as it has nothing in the input which could let it decide how to choose which type it should be. We know that we want float with float complex, but there doesn't appear to be a syntax to do this kind of 1-1 matching up.

Reply:
OK, when Travis finishes  will approve, assuming no errors